pkg/route/secret/secret_monitor.go

```diff
@@ -0,0 +1,215 @@
+package secret
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+
+	corev1 "k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+)
+
+type SecretEventHandlerRegistration interface {
+	cache.ResourceEventHandlerRegistration
+
+	GetKey() ObjectKey
+	GetHandler() cache.ResourceEventHandlerRegistration
+}
+
+type SecretMonitor interface {
+	//
+	AddSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error)
+	//
+	RemoveSecretEventHandler(SecretEventHandlerRegistration) error
+	//
+	GetSecret(SecretEventHandlerRegistration) (*v1.Secret, error)
+}
+
+type secretEventHandlerRegistration struct {
+	cache.ResourceEventHandlerRegistration
+	// objectKey will populate during AddEventHandler, and will be used during RemoveEventHandler
+	objectKey ObjectKey
+}
+
+func (r *secretEventHandlerRegistration) GetKey() ObjectKey {
+	return r.objectKey
+}
+
+func (r *secretEventHandlerRegistration) GetHandler() cache.ResourceEventHandlerRegistration {
+	return r.ResourceEventHandlerRegistration
+}
+
+type secretMonitor struct {
+	kubeClient kubernetes.Interface
+
+	lock sync.RWMutex
+	// monitors is map of singleItemMonitor. Each singleItemMonitor monitors/watches
+	// a secret through individual informer.
+	monitors map[ObjectKey]*singleItemMonitor
+}
+
+func NewSecretMonitor(kubeClient kubernetes.Interface) SecretMonitor {
+	return &secretMonitor{
+		kubeClient: kubeClient,
+		monitors:   map[ObjectKey]*singleItemMonitor{},
+	}
+}
+
+// create/update secret watch.
+// routeSecretName is a combination of "routename_secretname"
+func (s *secretMonitor) AddSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler) (SecretEventHandlerRegistration, error) {
+	return s.addSecretEventHandler(namespace, routeSecretName, handler, nil)
+}
+
+// addSecretEventHandler should only be used directly for tests. For production use AddSecretEventHandler().
+// createInformerFn helps in mocking sharedInformer for unit tests.
+func (s *secretMonitor) addSecretEventHandler(namespace, routeSecretName string, handler cache.ResourceEventHandler, createInformerFn func() cache.SharedInformer) (SecretEventHandlerRegistration, error) {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if !validateName(routeSecretName) {
+		return nil, fmt.Errorf("invalid routeSecretName combination : %s", routeSecretName)
+	}
+
+	// Inside a namespace multiple route can refer a common secret.
```

Comment:
multiple handlers for a single secret is fine.  I don't expect this code to need to know which route it is.

Reply:
Yeah agreed! I've tried to make `secret_manager` more generic and hence route is not required any more.  

This logic of multiple handlers check will be done inside `manager` now.

```go
	// each route (namespace/routeName) should be registered only once with any secret.
	// Note: inside a namespace multiple different routes can be registered(watch) with a common secret
	key := generateKey(namespace, routeName)
	if _, exists := m.registeredHandlers[key]; exists {
		return apierrors.NewInternalError(fmt.Errorf("route already registered with key %s", key))
	}
```
Let me know if this looks better.